editors/code/src/cargo.ts

```diff
@@ -0,0 +1,109 @@
+import * as cp from 'child_process';
+import * as readline from 'readline';
+import { OutputChannel } from 'vscode';
+
+interface CompilationArtifact {
+    fileName: string;
+    name: string;
+    kind: string;
+    isTest: boolean;
+}
+
+export class Cargo {
+    rootFolder: string;
+    env?: Record<string, string>;
+    output: OutputChannel;
+
+    public constructor(cargoTomlFolder: string, output: OutputChannel, env: Record<string, string> | undefined = undefined) {
+        this.rootFolder = cargoTomlFolder;
+        this.output = output;
+        this.env = env;
+    }
+
+    public async artifactsFromArgs(cargoArgs: string[]): Promise<CompilationArtifact[]> {
+        let artifacts: CompilationArtifact[] = [];
+
+        try {
+            await this.runCargo(cargoArgs,
+                message => {
+                    if (message.reason == 'compiler-artifact' && message.executable) {
+                        let isBinary = message.target.crate_types.includes('bin');
+                        let isBuildScript = message.target.kind.includes('custom-build');
+                        if ((isBinary && !isBuildScript) || message.profile.test) {
+                            artifacts.push({
+                                fileName: message.executable,
+                                name: message.target.name,
+                                kind: message.target.kind[0],
+                                isTest: message.profile.test
+                            })
+                        }
+                    }
+                    else if( message.reason == 'compiler-message') {
+                        this.output.append(message.message.rendered);
+                    }
+                },
+                stderr => {
+                    this.output.append(stderr);
+                }
+            );
+        }
+        catch (err) {
+            this.output.show(true);
+            throw new Error(`Cargo invocation has failed: ${err}`);
+        }
+
+        return artifacts;
+    }
+
+    public async executableFromArgs(cargoArgs: string[], extraArgs?: string[]): Promise<string> {
+        cargoArgs.push("--message-format=json");
+        if (extraArgs) {
+            cargoArgs.push('--');
+            cargoArgs.push(...extraArgs);
+        }
```

Comment:
I think we should not be pushing `extraArgs` here. `extraArgs` are for the final executable. 

IE, if we are running a single test, the args would be `cargo test --test foo` and the extra args `-- name::of::the::test`. Given that you already pass extra args to exetucable, I think this can be just removed.

Reply:
@matklad you are right. I've missed this point. Thanks.